### PR TITLE
fix SDK6 heap dumper documentation to reflect actual usage.

### PR DESCRIPTION
**Pull request by tboue**
Documentation didn't reflect what we had to do with @dalgassimou to perform a heap dump on device with architecture 8.

There also is a duplicate of this page for SDK5, do I update it ?

### CODE REVIEW

File: SDK6UserGuide/heapAnalyzer.rst [hunk 1]
```diff
@@ -75,11 +75,11 @@ The Heap Analyzer is an Eclipse IDE plugin that adds three tools to the MicroEJ
...
 
-When the Heap Dumper option is enabled, 
-the garbage collector process ends by performing a dump that represents a snapshot of the heap at this moment. 
-To generate such dump, you must explicitly call the ``System.gc()`` method in your code.
+The Heap Dumper generates heap files either :
+- directly from the Java heap on simulator 
+- from .hex files that must be manually retrieved on device.
 
-This tool can be used with the Simulator and with a device.
+The heap dump should be performed after a call to ``System.gc()`` to exclude discardable objects.
```
Comment (dalgassimou) [inline SDK6UserGuide/heapAnalyzer.rst:82]: the dump can be generation in many cases not only after a `System.gc` see here https://github.com/MicroEJ/docs/blob/master/ApplicationDeveloperGuide/debuggerProxy.rst?plain=1#L23

File: SDK6UserGuide/heapAnalyzer.rst [hunk 2]
```diff
@@ -75,11 +75,11 @@ The Heap Analyzer is an Eclipse IDE plugin that adds three tools to the MicroEJ
 Heap Dumper
 -----------
 
-When the Heap Dumper option is enabled, 
-the garbage collector process ends by performing a dump that represents a snapshot of the heap at this moment. 
-To generate such dump, you must explicitly call the ``System.gc()`` method in your code.
+The Heap Dumper generates heap files either :
+- directly from the Java heap on simulator 
+- from .hex files that must be manually retrieved on device.
```
Comment (dalgassimou) [inline SDK6UserGuide/heapAnalyzer.rst:80]:
```suggestion
- from .hex files that must be manually retrieved from the device.
```

File: SDK6UserGuide/heapAnalyzer.rst [hunk 3]
```diff
@@ -101,8 +101,17 @@ In order to generate a Heap dump of an Application running on a device:
...
 
-When the Application is executed on the device and the ``System.gc()`` method is called, 
-a Heap dump is performed and stored in the device memory.
+When the Application is executed on the device, the method ``System.gc()`` ends by calling ``LLMJVM_on_Runtime_gc_done`` callback.
+This method isn't always implemented as it is mostly used to perform a heap dump right after the garbage collection.
+If the symbol isn't embedded, add the following lines in any source file (ex:main.c):
+
+.. code:: c
+
+   void LLMJVM_on_OutOfMemoryError_thrown(){
```
Comment (dalgassimou) [inline SDK6UserGuide/heapAnalyzer.rst:110]: this is the wrong example, you should define here the LLMJVM_on_Runtime_gc_done

**Comment by dalgassimou** [comment]
We should add a section in the documentation, listing the memory sections to be emptied depending on whether we're in a multi or mono use case.

**Comment by tboue** [comment]
> We should add a section in the documentation, listing the memory sections to be emptied depending on whether we're in a multi or mono use case.

I agree but is it possible ? We did dump the entire RAM at the end because finding the list of sections was too long ?

File: SDK6UserGuide/heapAnalyzer.rst (hunk 2 above)
Comment (tboue) [reply SDK6UserGuide/heapAnalyzer.rst:80]: applied in https://github.com/MicroEJ/docs/pull/884/commits/437e9770e60d925051a25d490677e7d73069f934

File: SDK6UserGuide/heapAnalyzer.rst (hunk 1 above)
Comment (tboue) [reply SDK6UserGuide/heapAnalyzer.rst:82]: Explanation added in https://github.com/MicroEJ/docs/pull/884/commits/9272f8c920fe424effb96bb4b13a4af7ae5bc4ec

File: SDK6UserGuide/heapAnalyzer.rst (hunk 3 above)
Comment (tboue) [reply SDK6UserGuide/heapAnalyzer.rst:119]: fixed in https://github.com/MicroEJ/docs/pull/884/commits/9272f8c920fe424effb96bb4b13a4af7ae5bc4ec

File: SDK6UserGuide/heapAnalyzer.rst [hunk 4]
```diff
@@ -129,6 +147,15 @@ Retrieving the ``.hex`` file from the device can be done with Eclipse CDT and GD
 
 You now have the ``.hex`` file and need to extract the Heap dump.
 
+When using the tool in a multi-application context, additionnal sections must be dumped:
+- The dynamic features table:
+
+   dump ihex memory java_features_dynamic_start java_features_dynamic_end
```
Comment (dalgassimou) [inline SDK6UserGuide/heapAnalyzer.rst:153]: Is this correct? or
```suggestion
   dump ihex memory &java_features_dynamic_start &java_features_dynamic_end
```

File: SDK6UserGuide/heapAnalyzer.rst [hunk 5]
```diff
@@ -129,6 +147,15 @@ Retrieving the ``.hex`` file from the device can be done with Eclipse CDT and GD
...
+
+- features related sections. These will change depending on the board, number of features and other parameters. It is possible to dump the entire memory and let the :ref:`heap extractor<_sdk6_heapdumper_extract_heap>` sort the requiered sections instead.
```
Comment (dalgassimou) [inline SDK6UserGuide/heapAnalyzer.rst:155]:
```suggestion
- features related sections. These will change depending on the board, number of features and other parameters. It is possible to dump the entire memory and let the :ref:`heap extractor<_sdk6_heapdumper_extract_heap>` extract the requiered sections instead.
```

File: SDK6UserGuide/heapAnalyzer.rst [hunk 6]
```diff
@@ -101,8 +101,26 @@ In order to generate a Heap dump of an Application running on a device:
 - Update your Application code to call the ``System.gc()`` method where you need a Heap dump.
 - Build the Executable and deploy it on the device.
 
-When the Application is executed on the device and the ``System.gc()`` method is called, 
-a Heap dump is performed and stored in the device memory.
+When the Application is executed on the device, the method ``System.gc()`` ends by calling ``LLMJVM_on_Runtime_gc_done`` callback.
+This method isn't always implemented as it is mostly used to perform a heap dump right after the garbage collection.
+If the symbol isn't embedded, add the following lines in any source file (ex:main.c):
```
Comment (dalgassimou) [inline SDK6UserGuide/heapAnalyzer.rst:106]: I don't think this is right, I purpose something like this
```suggestion
If the symbol is inlined, add the following lines in any source file (ex:main.c):
```
What do you think @frivieremicroej?

File: SDK6UserGuide/heapAnalyzer.rst [hunk 7]
```diff
@@ -75,60 +75,92 @@ The Heap Analyzer is an Eclipse IDE plugin that adds three tools to the MicroEJ
...
+
+   .. code:: c
+
+      void LLMJVM_on_Runtime_gc_done(){
+         //No need to add code to the function
+      }
+
+      void LLMJVM_on_OutOfMemoryError_thrown(){
+         //No need to add code to the function
+      }
```
Comment (tboue) [inline SDK6UserGuide/heapAnalyzer.rst:124]: Tested again, no need to implement the callback, just having a body header is enough

File: SDK6UserGuide/heapAnalyzer.rst (hunk 6 above)
Comment (tboue) [reply SDK6UserGuide/heapAnalyzer.rst]: Changed to https://github.com/MicroEJ/docs/pull/884/files#diff-d122d22cbbd8db54cff58c20f6831435524140fb2fafe25be7c8dcc084801a51R113-R114

**Review by frivieremicroej — CHANGES_REQUESTED** [review]

File: SDK6UserGuide/heapAnalyzer.rst [hunk 8]
```diff
@@ -75,60 +75,92 @@ The Heap Analyzer is an Eclipse IDE plugin that adds three tools to the MicroEJ
...
+  - :ref:`Extract the Heap dump from the hex file <sdk6_heapdumper_extract_heap>`
+
+.. note::
+
+   Core Engine hooks may have been inlined by the third-party linker. 
+   If the symbol is not accessible to the debugger, you must declare them in your VEE Port:
+
+   .. code:: c
+
+      void LLMJVM_on_Runtime_gc_done(){
```
[the remark this answers is not in the document]
Comment (frivieremicroej) [reply SDK6UserGuide/heapAnalyzer.rst:118]: confirmed by @tboue

**Review by frivieremicroej — APPROVED** [review]

**Pull request merged**